session/session.go

```diff
@@ -341,6 +354,51 @@ func (s *MenderShellSession) IsExpired(setStatus bool) bool {
 	return e
 }
 
+func (s *MenderShellSession) healthcheck() {
+	nextHealthcheckPing := time.Now().Add(healthcheckInterval)
+	s.healthcheckTimeout = time.Now().Add(healthcheckInterval + healthcheckTimeout)
+
+	for {
+		select {
+		case <-s.stop:
+			return
+		case <-time.After(time.Until(s.healthcheckTimeout)):
+			if s.healthcheckTimeout.Before(time.Now()) {
+				log.Errorf("session %s, health check failed, connection with the client lost", s.id)
+				s.expiresAt = time.Now()
+				return
+			}
+		case <-time.After(time.Until(nextHealthcheckPing)):
+			s.healthcheckPing()
+			nextHealthcheckPing = time.Now().Add(healthcheckInterval)
+		}
+	}
+}
+
+func (s *MenderShellSession) healthcheckPing() {
+	msg := &ws.ProtoMsg{
+		Header: ws.ProtoHdr{
+			Proto:     ws.ProtoTypeShell,
+			MsgType:   wsshell.MessageTypePingShell,
+			SessionID: s.id,
+			Properties: map[string]interface{}{
+				"timeout": int(healthcheckInterval.Seconds() + healthcheckTimeout.Seconds()),
+				"status":  wsshell.ControlMessage,
+			},
+		},
+		Body: nil,
+	}
+	log.Debugf("session %s healthcheck ping", s.id)
+	err := connectionmanager.Write(ws.ProtoTypeShell, msg)
+	if err != nil {
+		log.Debugf("error on write: %s", err.Error())
+	}
+}
+
+func (s *MenderShellSession) HealthcheckPong() {
+	s.healthcheckTimeout = time.Now().Add(healthcheckInterval + healthcheckTimeout)
```

Comment:
this write happens in the daemon main loop; different thread than [this write](https://github.com/mendersoftware/mender-connect/pull/9/commits/17adb2019cc6028d858f364b9b05bde987599c4e#diff-10dab83c04a99358d431906d37f48b92be97d50a2719d4ddf81c62d3b03a0aebR359) and [these reads](https://github.com/mendersoftware/mender-connect/pull/9/commits/17adb2019cc6028d858f364b9b05bde987599c4e#diff-10dab83c04a99358d431906d37f48b92be97d50a2719d4ddf81c62d3b03a0aebR365) do you think we could use a mutex for protection?

Another thing (somewhat nitpicking): if I run a session on Mar 28th 2021 at 01:59 CEST or on Oct 31st 2021 at 02:59 CEST when DST start/end take place. will it mean my session will be in trouble?

Reply:
> this write happens in the daemon main loop; different thread than [this write](https://github.com/mendersoftware/mender-connect/pull/9/commits/17adb2019cc6028d858f364b9b05bde987599c4e#diff-10dab83c04a99358d431906d37f48b92be97d50a2719d4ddf81c62d3b03a0aebR359) and [these reads](https://github.com/mendersoftware/mender-connect/pull/9/commits/17adb2019cc6028d858f364b9b05bde987599c4e#diff-10dab83c04a99358d431906d37f48b92be97d50a2719d4ddf81c62d3b03a0aebR365) do you think we could use a mutex for protection?

Well spotted. Due to the long timers between receiving a pong and triggering the next ping, triggering a race condition is extremely unlikely, but I still consider this somewhat bad style, and it may have more serious consequences later if this is changed. My suggested fix is: Instead of updating `s.healthcheckTimeout` here, use a channel to send the value back to the ping loop, and let it update the variable in there instead. Then all accesses are from the same thread.

Reply:
> Another thing (somewhat nitpicking): if I run a session on Mar 28th 2021 at 01:59 CEST or on Oct 31st 2021 at 02:59 CEST when DST start/end take place. will it mean my session will be in trouble?

This should be fine, the Golang docs state that all "time differences" (as opposed to absolute times) are timezone independent.

Reply:
Addressed here: https://github.com/mendersoftware/mender-connect/pull/9/commits/3d7bfc5baba385941b7c64df50adcced6ff899c9